### PR TITLE
Save and Restore State with GV 🦎 Closes #2174 

### LINKED ISSUE #2174: GeckoView: Save and restore state (when switching tabs)
There are certain situations (like switching tabs) where we save the `IWebView` state in the session object and restore it later. This is not implemented for GeckoView yet. In this situation we are reloading the last URL and have lost the whole history of that session.

### PR DESCRIPTION
**Pull request by ekager**

### CODE REVIEW

File: app/src/gecko/java/org/mozilla/focus/web/WebViewProvider.java [hunk 1]
```diff
@@ -467,12 +471,36 @@ public boolean canGoBack() {
...
+                geckoSession.restoreState(sessionState);
+            } else {
+                loadUrl(desiredURL);
+            }
         }
 
         @Override
-        public void saveWebViewState(@NonNull Session session) {
-            // TODO: save anything needed for navigation history restoration.
+        public void saveWebViewState(@NonNull final Session session, final CountDownLatch latch) {
```
Comment (pocmo) [inline app/src/gecko/java/org/mozilla/focus/web/WebViewProvider.java]: Let's not expose the CountDownLatch in the API on `IWebView`. You should be able to create the latch here, post the runnable and then wait .. all inside saveWebViewState here?

File: app/src/main/java/org/mozilla/focus/fragment/WebFragment.java [hunk 2]
```diff
@@ -104,7 +106,17 @@ public void applyLocale() {
...
+                final CountDownLatch latch = new CountDownLatch(1);
+                webView.saveWebViewState(session, latch);
+                try {
+                    latch.await();
+                } catch (InterruptedException e) {
+                    // Do nothing, the page was not saved
+                }
+            } else {
+                webView.saveWebViewState(session, null);
+            }
```
Comment (pocmo) [inline app/src/main/java/org/mozilla/focus/fragment/WebFragment.java]: As mentioned above: If you hide the CountDownLatch inside saveWebViewState() then this is just a single call here again.

**Comment by ekager** [comment]
Makes sense :)

File: app/src/gecko/java/org/mozilla/focus/web/WebViewProvider.java [hunk 3]
```diff
@@ -467,12 +471,42 @@ public boolean canGoBack() {
...
+                geckoSession.restoreState(sessionState);
+            } else {
+                loadUrl(desiredURL);
+            }
         }
 
         @Override
-        public void saveWebViewState(@NonNull Session session) {
-            // TODO: save anything needed for navigation history restoration.
+        public void saveWebViewState(@NonNull final Session session) {
```
Comment (sputnikci) [inline app/src/gecko/java/org/mozilla/focus/web/WebViewProvider.java]: \[PMD\] INFO: The method 'saveWebViewState\(Session\)' has a NCSS line count of 14\.

File: app/src/gecko/java/org/mozilla/focus/web/WebViewProvider.java [hunk 4]
```diff
@@ -467,12 +471,43 @@ public boolean canGoBack() {
...
+            } else {
+                loadUrl(desiredURL);
+            }
         }
 
         @Override
-        public void saveWebViewState(@NonNull Session session) {
-            // TODO: save anything needed for navigation history restoration.
+        @SuppressWarnings({"PMD.NcssMethodCount"})
+        public void saveWebViewState(@NonNull final Session session) {
```
Comment (sputnikci) [inline app/src/gecko/java/org/mozilla/focus/web/WebViewProvider.java]: \[PMD\] INFO: The method 'saveWebViewState\(Session\)' has a NCSS line count of 14\.

**Pull request merged**